File: sdk/python/myzubster.py
"""
MyZubster Gateway Python SDK
A Python client library for the MyZubster Gateway API.

Installation:
    pip install myzubster

Usage:
    from myzubster import MyZubsterClient
    
    client = MyZubsterClient()
    client.login("username", "password")
    client.register_animal(name="Buddy", animal_type="dog")
"""
import requests
from typing import Optional, Dict, Any


class MyZubsterError(Exception):
    """Exception raised for MyZubster API errors."""
    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        self.message = message
        super().__init__(f"[{status_code}] {message}")


class MyZubsterClient:
    """Client for the MyZubster Gateway API."""
    
    def __init__(self, base_url: str = "https://myzubsterapp.onrender.com"):
        self.base_url = base_url.rstrip("/")
        self.token: Optional[str] = None
        self.session = requests.Session()
    
    def _request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        url = f"{self.base_url}{endpoint}"
        if self.token:
            self.session.headers.update({"Authorization": f"Bearer {self.token}"})
        response = self.session.request(method, url, **kwargs)
        data = response.json() if response.content else {}
        if not response.ok:
            raise MyZubsterError(response.status_code, data.get("error", response.text))
        return data
    
    def login(self, username: str, password: str) -> Dict:
        """Authenticate and store JWT token."""
        data = self._request("POST", "/api/auth/login", json={"username": username, "password": password})
        self.token = data.get("token")
        return data
    
    def register(self, username: str, password: str, email: str) -> Dict:
        """Register a new user."""
        return self._request("POST", "/api/auth/register", json={"username": username, "password": password, "email": email})
    
    def get_profile(self) -> Dict:
        """Get current user profile."""
        return self._request("GET", "/api/auth/profile")
    
    def get_health(self) -> Dict:
        """Check API health status."""
        return self._request("GET", "/api/health")
    
    def get_info(self) -> Dict:
        """Get gateway information."""
        return self._request("GET", "/api/info")
    
    def register_animal(self, name: str, animal_type: str, **extra) -> Dict:
        """Register a new animal."""
        data = {"name": name, "type": animal_type, **extra}
        return self._request("POST", "/api/animals/register", json=data)
    
    def list_animals(self) -> Dict:
        """List all animals."""
        return self._request("GET", "/api/animals")
    
    def get_animal(self, animal_id: str) -> Dict:
        """Get animal details."""
        return self._request("GET", f"/api/animals/{animal_id}")
    
    def register_plant(self, name: str, plant_type: str, **extra) -> Dict:
        """Register a new plant."""
        data = {"name": name, "type": plant_type, **extra}
        return self._request("POST", "/api/plants/register", json=data)
    
    def list_plants(self) -> Dict:
        """List all plants."""
        return self._request("GET", "/api/plants")
    
    def create_robot(self, robot_id: str, name: str, wallet_address: str) -> Dict:
        """Create a new robot."""
        return self._request("POST", "/api/robot/create", json={"robotId": robot_id, "name": name, "walletAddress": wallet_address})
    
    def assign_job(self, robot_id: str, job_id: str, client_id: str, amount: float, currency: str = "MYZ") -> Dict:
        """Assign a job to a robot."""
        return self._request("POST", "/api/robot/assign", json={"robotId": robot_id, "jobId": job_id, "clientId": client_id, "amount": amount, "currency": currency})
    
    def complete_job(self, robot_id: str, job_id: str, result: str) -> Dict:
        """Mark a robot job as complete."""
        return self._request("POST", "/api/robot/job/complete", json={"robotId": robot_id, "jobId": job_id, "result": result})
    
    def get_swap_rate(self, from_currency: str, to_currency: str, amount: float) -> Dict:
        """Get XMR/MYZ swap rate."""
        return self._request("GET", "/api/swap/rate", params={"from": from_currency, "to": to_currency, "amount": amount})
    
    def execute_swap(self, from_currency: str, to_currency: str, amount: float, user_id: str) -> Dict:
        """Execute a currency swap."""
        return self._request("POST", "/api/swap/execute", json={"from": from_currency, "to": to_currency, "amount": amount, "userId": user_id})
    
    def get_bounties(self) -> Dict:
        """List all open bounties."""
        return self._request("GET", "/api/bounties")
    
    def get_rewards(self, user_id: str) -> Dict:
        """Get reward history for a user."""
        return self._request("GET", "/api/rewards", params={"userId": user_id})

File: sdk/go/myzubster.go
// Package myzubster provides a Go client for the MyZubster Gateway API.
//
// Installation:
//   go get github.com/MyZubster-Ecosystem/MyZubsterGateway/sdk/go
//
// Usage:
//   client := myzubster.NewClient("https://myzubsterapp.onrender.com")
//   client.Login("username", "password")
//   client.RegisterAnimal("Buddy", "dog")
package myzubster

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// Client represents a MyZubster Gateway API client.
type Client struct {
	BaseURL string
	Token   string
	HTTP    *http.Client
}

// NewClient creates a new MyZubster client.
func NewClient(baseURL string) *Client {
	if baseURL == "" {
		baseURL = "https://myzubsterapp.onrender.com"
	}
	return &Client{
		BaseURL: strings.TrimSuffix(baseURL, "/"),
		HTTP:    http.DefaultClient,
	}
}

func (c *Client) doRequest(method, endpoint string, body interface{}, params map[string]string) (map[string]interface{}, error) {
	u := c.BaseURL + endpoint
	if len(params) > 0 {
		v := url.Values{}
		for k, val := range params {
			v.Set(k, val)
		}
		u += "?" + v.Encode()
	}

	var bodyReader io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		bodyReader = bytes.NewReader(jsonData)
	}

	req, err := http.NewRequest(method, u, bodyReader)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.Token)
	}

	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	if resp.StatusCode >= 400 {
		return result, fmt.Errorf("[%d] %v", resp.StatusCode, result["error"])
	}
	return result, nil
}

// Login authenticates and stores the JWT token.
func (c *Client) Login(username, password string) (map[string]interface{}, error) {
	data, err := c.doRequest("POST", "/api/auth/login", map[string]string{
		"username": username,
		"password": password,
	}, nil)
	if err != nil {
		return data, err
	}
	if token, ok := data["token"].(string); ok {
		c.Token = token
	}
	return data, nil
}

// Register registers a new user.
func (c *Client) Register(username, password, email string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/auth/register", map[string]string{
		"username": username, "password": password, "email": email,
	}, nil)
}

// GetProfile returns the current user profile.
func (c *Client) GetProfile() (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/auth/profile", nil, nil)
}

// GetHealth checks API health status.
func (c *Client) GetHealth() (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/health", nil, nil)
}

// GetInfo returns gateway information.
func (c *Client) GetInfo() (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/info", nil, nil)
}

// RegisterAnimal registers a new animal.
func (c *Client) RegisterAnimal(name, animalType string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/animals/register", map[string]string{
		"name": name, "type": animalType,
	}, nil)
}

// ListAnimals lists all animals.
func (c *Client) ListAnimals() (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/animals", nil, nil)
}

// RegisterPlant registers a new plant.
func (c *Client) RegisterPlant(name, plantType string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/plants/register", map[string]string{
		"name": name, "type": plantType,
	}, nil)
}

// CreateRobot creates a new robot.
func (c *Client) CreateRobot(robotID, name, walletAddress string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/robot/create", map[string]string{
		"robotId": robotID, "name": name, "walletAddress": walletAddress,
	}, nil)
}

// AssignJob assigns a job to a robot.
func (c *Client) AssignJob(robotID, jobID, clientID string, amount float64, currency string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/robot/assign", map[string]interface{}{
		"robotId": robotID, "jobId": jobID, "clientId": clientID,
		"amount": amount, "currency": currency,
	}, nil)
}

// CompleteJob marks a robot job as complete.
func (c *Client) CompleteJob(robotID, jobID, result string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/robot/job/complete", map[string]string{
		"robotId": robotID, "jobId": jobID, "result": result,
	}, nil)
}

// GetSwapRate gets XMR/MYZ swap rate.
func (c *Client) GetSwapRate(from, to string, amount float64) (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/swap/rate", nil, map[string]string{
		"from": from, "to": to, "amount": fmt.Sprintf("%v", amount),
	})
}

// ExecuteSwap executes a currency swap.
func (c *Client) ExecuteSwap(from, to string, amount float64, userID string) (map[string]interface{}, error) {
	return c.doRequest("POST", "/api/swap/execute", map[string]interface{}{
		"from": from, "to": to, "amount": amount, "userId": userID,
	}, nil)
}

// GetBounties lists all open bounties.
func (c *Client) GetBounties() (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/bounties", nil, nil)
}

// GetRewards gets reward history for a user.
func (c *Client) GetRewards(userID string) (map[string]interface{}, error) {
	return c.doRequest("GET", "/api/rewards", nil, map[string]string{
		"userId": userID,
	})
}
